### ISSUE TITLE
Printing constant division by zero causes compiler panic

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust#37998]
Trying to print constant that is division by zero causes Rust compiler panic

I tried this code:

```rust
const WHY: i32 = 1 / 0;

fn main() {
    println!("{}", WHY);
}
```

I expected to see this happen: Just an error

Instead, this happened:

```
warning: constant evaluation error: attempt to divide by zero. This will become a HARD ERROR in the future, #[warn(const_err)] on by default
 --> x.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^

error[E0080]: constant evaluation error
 --> x.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^ attempt to divide by zero

error: internal compiler error: src/librustc_trans/mir/constant.rs:392: tmp0 not initialized
 --> x.rs:4:20
  |
4 |     println!("{}", WHY);
  |                    ^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:602
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

## Meta

`rustc --version --verbose`:

```
rustc 1.13.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-freebsd
release: 1.13.0
```

Backtrace:

(Rust on FreeBSD installed from repository doesn't show a backtrace, and I'm not in home right now, this stacktrace comes from playpen)

```
thread 'rustc' panicked at 'Box<Any>', ../src/librustc_errors/lib.rs:602
stack backtrace:
   1:     0x7f4ec53b785f - std::sys::backtrace::tracing::imp::write::h6f1d53a70916b90d
   2:     0x7f4ec53c6dbd - std::panicking::default_hook::{{closure}}::h137e876f7d3b5850
   3:     0x7f4ec53c422e - std::panicking::default_hook::h0ac3811ec7cee78c
   4:     0x7f4ec53c4918 - std::panicking::rust_panic_with_hook::hc303199e04562edf
   5:     0x7f4ec3efb378 - std::panicking::begin_panic::hf3c77cb5fd5814c4
   6:     0x7f4ec3ff0df5 - rustc::session::opt_span_bug_fmt::{{closure}}::hf0e25c590c479ec5
   7:     0x7f4ec3f1985a - rustc::session::span_bug_fmt::h42d4c0a219d5e4ec
   8:     0x7f4ec3fcfe36 - rustc_trans::mir::constant::MirConstContext::const_lvalue::h19c47a857fb60d91
   9:     0x7f4ec3fd1620 - rustc_trans::mir::constant::MirConstContext::const_rvalue::h8c7c680bb5d720f3
  10:     0x7f4ec3fce08b - rustc_trans::mir::constant::MirConstContext::trans::h96c052b8d8423eda
  11:     0x7f4ec3fd512c - rustc_trans::mir::constant::<impl rustc_trans::mir::MirContext<'bcx, 'tcx>>::trans_constant::hdfac2835d2e37417
  12:     0x7f4ec3fd6dca - rustc_trans::mir::operand::<impl rustc_trans::mir::MirContext<'bcx, 'tcx>>::trans_operand::h1937d4fb09193ffc
  13:     0x7f4ec3fda22c - rustc_trans::mir::rvalue::<impl rustc_trans::mir::MirContext<'bcx, 'tcx>>::trans_rvalue_operand::ha8ed8ebcb69127d4
  14:     0x7f4ec3fc6264 - rustc_trans::mir::block::<impl rustc_trans::mir::MirContext<'bcx, 'tcx>>::trans_block::hadeb870961b3c139
  15:     0x7f4ec3fc4168 - rustc_trans::mir::trans_mir::h5581e064c4dd5b26
  16:     0x7f4ec3f61d2c - rustc_trans::base::trans_closure::h62865eeceeb9846e
  17:     0x7f4ec3fe3cc3 - rustc_trans::trans_item::TransItem::define::h2b8862ddc7ac3e96
  18:     0x7f4ec3f65414 - rustc_trans::base::trans_crate::h7b6c2eb43ae0c1da
  19:     0x7f4ec5748a0d - rustc_driver::driver::phase_4_translate_to_llvm::h0646ffc4a1716d29
  20:     0x7f4ec5783e2a - rustc_driver::driver::compile_input::{{closure}}::h5a81b1e8a8f3f531
  21:     0x7f4ec576f6a3 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h4091eeab6f271a29
  22:     0x7f4ec57463ad - rustc_driver::driver::phase_3_run_analysis_passes::hff1c31d493acfd09
  23:     0x7f4ec5732a39 - rustc_driver::driver::compile_input::hef221f9f618e92f5
  24:     0x7f4ec575d206 - rustc_driver::run_compiler::h7dddd4089f8330b4
  25:     0x7f4ec56992c3 - std::panicking::try::do_call::h1e3ece12e3f812df
  26:     0x7f4ec53cecb6 - __rust_maybe_catch_panic
  27:     0x7f4ec56b74f9 - <F as alloc::boxed::FnBox<A>>::call_box::h05b222e46803eb30
  28:     0x7f4ec53c2c80 - std::sys::thread::Thread::new::thread_start::h759e10bc4abc7e72
  29:     0x7f4ebe30c709 - start_thread
  30:     0x7f4ec508682c - clone
  31:                0x0 - <unknown>
```

**Comment by oli-obk** [comment]
should probably turn this warning into an error.

**Comment by aochagavia** [comment]
As you can see on the [blame](https://github.com/rust-lang/rust/blame/e804a3cf256106c097d44f6e0212cd183122da07/src/librustc_passes/consts.rs#L117) the warning has been there for 10 months. It seems like a good moment to turn it into a hard error.

**Comment by aochagavia** [comment]
@oli-obk I have a branch where I use `span_err!` to indicate an error in the evaluation of the constant. However, I need to specify an error code... Is there already an error code for this or do I need to come up with one?

**Comment by oli-obk** [comment]
@aochagavia there is the https://doc.rust-lang.org/error-index.html#E0020 error

**Comment by aochagavia** [comment]
@oli-obk Thanks! I didn't know of the error index. I found https://doc.rust-lang.org/error-index.html#E0080 which seems better since the code should handle other types of error as well.

**Comment by xfix** [comment]
This error also happens for other constant evaluation errors.

```rust
const WHY: u32 = 0 - 1;

fn main() {
    println!("{}", WHY);
}
```

**Comment by cyplo** [comment]
The error seems to be 
```
warning: constant evaluation error: attempt to divide by zero. This will become a HARD ERROR in the future
 --> main.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> main.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^ attempt to divide by zero

error: aborting due to previous error
``` on rustc 1.22.0-nightly (4c053db23 2017-10-22)

Would that be enough to close this issue as no ICE anymore ?

**Comment by xfix** [comment]
The error message is less than great, but I suppose it's not ICE anymore.

**Comment by oli-obk** [comment]
The error message now is

```
warning: constant evaluation error: attempt to divide by zero
 --> src/main.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> src/main.rs:1:18
  |
1 | const WHY: i32 = 1 / 0;
  |                  ^^^^^ attempt to divide by zero

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```
